testing new tasks

carlmala opened the pull request:
added new tasks as a test,
added sissyDressup, which generates a tasklist out of random selected items,
added sissyDressup to generateAction, so that dressup is called first before other tasks, but I don't think, that's the right implementation.
quite sloppy at the moment

thefapinstructor commented:
Thanks for the pull request!  I see a bunch of great ideas.  It looks like you are still working on this branch since I see some debugging code and all of the original tasks are disabled by default.   Once you are ready throw me another PR for code review and merge into master.

carlmala commented:
Alright, restored the default selection and added some new ones and I edited the probabilities of the new tasks.
So should be fine, should be working now.

thefapinstructor commented on src/game/actions/anal/femCondom.js:
```diff
@@ -0,0 +1,29 @@
+    return [done];
+  }
+};
+insertFemCondom.label = "Insert female condom in your ass";
```
Should be a title.  Renamed to 'Female Condom'

thefapinstructor commented on src/game/actions/anal/analDildo.js:
```diff
@@ -0,0 +1,51 @@
+    return [done];
+  }
+};
+removeAnalDildo.label = "Remove Analdildo";
```
rename label to "Remove Anal Dildo"

carlmala commented on src/game/actions/generateAction.js:
```diff
@@ -45,9 +46,15 @@ const generateAction = () => {
-    // get one of the chosen actions
-    action =
-      chosenActions[getRandomInclusiveInteger(0, chosenActions.length - 1)];
+    if (dressupAction === false && store.config.tasks.sissydressup === true) {
```
The idea was to call the dressup action as the first action in game if the user activates it, because I thought, it would make more sense than activate the dressup action mid-game. What do you think?

thefapinstructor replied:
Gotcha, thanks for the explanation.  That would be a new concept, actions before the randomized actions begin.  Maybe for now make it random like the other actions with a 30% probability, that will trigger within probably the first minute or two.  I'll think of a more generic way to implement the pregame actions.

carlmala commented:
Alright, added some changes again. I have to thank you, it's quite a big project seeing all that code. 
I hope this time it's better, never did anything collaborative on github, so I'm still learning. 
I added a comment about the dressup topic above. 
And there's one question remaining: I changed to the buttInUse flag, which I think is a good idea. Now it would be nice to be able to change between the different anal tasks, but I'm not sure how to implement, also so that there's no chance that the task is changing after a few seconds again.
There should be the possibility that anal tasks can change, but also there should be a minimum delay between two different tasks...

thefapinstructor commented:
I accidentally approved the code review (almost there though).

- Please see latest comment on src/game/actions/generateAction.js 
- It looks like the formatting of the code got changed in ConfigPage leading to a huge amount of code differences.  I'm not sure what editor you are using but please install Prettier and run it on that file.  It will organize it the proper way.  That will help a lot looking for differences.

thefapinstructor commented on src/game/actions/dressup/sissyDressup.js:
```diff
@@ -0,0 +1,62 @@
+  dismissNotification
+} from "engine/createNotification";
+import { strokerRemoteControl } from "game/loops/strokerLoop";
+let ouvert = Math.round(Math.random());
```
I'm not sure of Math.round on a math.random makes sense.  Math.random returns a value between 0-0.9.  Round will round the value to either 0 or 1.   If you're intention was to have a random boolean use utils/math.getRandomBoolean().

thefapinstructor commented on src/game/actions/dressup/sissyDressup.js:
```diff
@@ -0,0 +1,62 @@
+} from "engine/createNotification";
+import { strokerRemoteControl } from "game/loops/strokerLoop";
+let ouvert = Math.round(Math.random());
+let chTop = ["", "Body", "Corsage", "BH"];
```
What does the prefix ch represent?  Usually I try to avoid adding prefix to variables names as a standard.  Maybe we can organize this list of variables a little more effectively.   Please read below and use as a reference, I hope it helps.

From below it looks like you want to create a list of clothing items for the user to put on.  I would recommend using a 2D array to achieve this.
```
const clothing = [
  ["Slip", "Panties"], 
  ["Nylons"],
  ["Pumps"]
];

// should return 1D array of clothing items, one from each group.
const putoncloths = clothing.map(group => group[Math.floor(Math.random() * group.length))

// Then use join to merge those values with a comma delimiter 
putoncloths.join(", ")
```

carlmala commented:
Alright, added some changes again. Code Formatting should now be the same (running prettier). 
I like the cleaner style of a 2d array, good idea! But I'm not sure how to implement some conditions. For example, Ouvert + Nylons doesn't make sense. How could I implement this? And also Body + Slip is a combination, that shouldn't occur.

thefapinstructor commented:
I can think of some ways to solve the combination problem. 

1. Group any incompatible pieces together, ["", "Ouvert", "Nylons"] for an example.
2. For more complex combinations you can create preconfigured outfits.
```
outfits = [
 ["Nylons", "Dress", "Pumps"],
 ["Nylons", "Dress", "HighHeels"]
]
```
3. The third way and probably most correct yet by far more technically difficult is to implement a tree like structure to map out clothing iterations.  Kind of like a dependency tree but for compatible items.

carlmala commented:
alright, added an object with the clothing items and groups. Now items will be selected from the object and then sorted to a result. But I think it's still not right. I think it could be a solution to create an object with items in it with different conditions. But I don't know an efficient solution to filter then the result-object out of the conditions.

thefapinstructor commented:
Yeah it's not an easy solution.  Let me know once it's working and I'll merge your PR into master.   Please merge from upstream master into your fork as there are now conflicts since I've been pushing new updates the past couple of days.

carlmala commented:
Ok, check the new version. Upstream is merged, so there should be no conflicts. The dressup cloth item generation is now a new version. 
Changed if to an object, where different variations are generated via groups. Then I select from the cloth items randomly, but now with a different distribution. `Math.pow(Math.random(),2)` should generate a weighted distribution like this:

> Probability graph of square of a random number
> ##################################################
> #####################
> ################
> ##############
> ############
> ###########
> ##########
> ##########
> #########
> #########
> ########
> ########
> ########
> #######
> #######
> #######
> #######
> ######
> ######
> ######
> ######
> ######
> ######
> ######
> ######
> #####
> #####
> #####
> #####
> #####

That should help reducing the combined clothing items because most of the time a lower value is generated which increases the probability that no item from a group is selected.
Then the results get filtered. And if there's no clothing item selected, which can happen with the random numbers, then there's a fallback.
